File: androidApp/src/main/kotlin/com/github/jvsena42/loopky/ui/components/PubkyAppLink.kt
package com.github.jvsena42.loopky.ui.components

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.OpenInNew
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.role
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.github.jvsena42.loopky.R
import com.github.jvsena42.loopky.ui.theme.LoopkyTheme

/**
 * The two ways Loopky points at pubky.app.
 *
 * A Loopky account *is* a Pubky account — the same `profile.json`, the same follow graph, the same
 * key — and until now nothing in the app said so out loud. Both of these are deliberately quiet:
 * the network underneath is worth knowing about, but it is not what someone opened a flashcards
 * app to do.
 *
 * Neither builds its own URL. The address comes from the ViewModel, which reads it off
 * `PubkyEnvironment` — a debug build points at staging, where its account actually exists.
 */

/**
 * The button that leaves for pubky.app: the mark alone, in the same outlined 48dp circle Copy and
 * Share wear, so it carries no more weight in the row than they do.
 *
 * Monochrome, in Share's grey. pubky.app sets the mark in lime on black, but that disc was the
 * highest-contrast thing on a cream screen and pulled the eye before the primary action beside it
 * — and the lime without it is the faintest thing on the screen. The shape alone says whose logo
 * this is, so it wears the same grey as the icons it sits with.
 */
@Composable
fun PubkyAppIconButton(
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val colors = LoopkyTheme.colors
    val label = stringResource(R.string.pubky_app_open_profile)

    Box(
        modifier = modifier
            .size(48.dp)
            .clip(CircleShape)
            .background(colors.surfaceCard)
            .border(1.dp, colors.borderSubtle, CircleShape)
            .clickable(onClick = onClick)
            // The mark carries the meaning visually and has no text of its own, so the label goes
            // on the button rather than on the image inside it.
            .semantics {
                role = Role.Button
                contentDescription = label
            },
        contentAlignment = Alignment.Center,
    ) {
        PubkyMark(size = 22.dp)
    }
}

/**
 * The self-profile call to action: one soft row explaining what the button above it does, for the
 * person who has no reason to know that the key they signed in with is also a social account.
 *
 * A card rather than a banner, and it never claims a Loopky deck appears there — it does not. What
 * travels is the profile and, when they choose to announce one, the post.
 */
@Composable
fun PubkyAppProfileCta(
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val colors = LoopkyTheme.colors

    Row(
        modifier = modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(20.dp))
            .background(colors.surfaceSecondary)
            .clickable(onClick = onClick)
            .testTag("profile_pubky_app_cta")
            .padding(16.dp),
        horizontalArrangement = Arrangement.spacedBy(14.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        // The same mark the button wears, on the card's own surface.
        PubkyMark(size = 26.dp)

        Column(
            modifier = Modifier.weight(1f),
            verticalArrangement = Arrangement.spacedBy(2.dp),
        ) {
            Text(
                text = stringResource(R.string.pubky_app_cta_title),
                fontSize = 14.sp,
                fontWeight = FontWeight.W700,
                color = colors.foregroundPrimary,
            )
            Text(
                text = stringResource(R.string.pubky_app_cta_body),
                fontSize = 12.sp,
                lineHeight = 17.sp,
                color = colors.foregroundMuted,
            )
        }

        Icon(
            imageVector = Icons.AutoMirrored.Filled.OpenInNew,
            contentDescription = null,
            tint = colors.foregroundMuted,
            modifier = Modifier.size(18.dp),
        )
    }
}

/**
 * pubky.app's mark, at [size], tinted like every other icon on the screen.
 */
@Composable
private fun PubkyMark(size: Dp) {
    Icon(
        painter = painterResource(R.drawable.ic_pubky),
        contentDescription = null,
        tint = LoopkyTheme.colors.foregroundSecondary,
        modifier = Modifier.size(size),
    )
}

@Preview
@Composable
private fun PubkyAppLinkPreview() {
    LoopkyTheme {
        Column(
            modifier = Modifier
                .background(LoopkyTheme.colors.surfacePrimary)
                .padding(16.dp),
            verticalArrangement = Arrangement.spacedBy(16.dp),
        ) {
            PubkyAppIconButton(onClick = {})
            PubkyAppProfileCta(onClick = {})
        }
    }
}

File: iosApp/iosApp/Views/PubkyAppLink.swift
import SwiftUI
import UIKit

/// The two ways Loopky points at pubky.app.
///
/// A Loopky account *is* a Pubky account — the same `profile.json`, the same follow graph, the
/// same key — and nothing on iOS said so: a foreign profile offered a generic globe, and the self
/// profile a plain "Open on pubky.app" button with no explanation of what that even is.
///
/// Both are deliberately quiet. The network underneath is worth knowing about, but it is not what
/// someone opened a flashcards app to do. Neither builds its own URL: the address comes from the
/// ViewModel, which reads it off `PubkyEnvironment`, so a debug build points at the staging
/// instance where its account actually exists.

/// pubky.app's mark, tinted like every other icon on the screen.
///
/// Monochrome on purpose. pubky.app sets the mark in lime on black, but that disc was the
/// highest-contrast thing on a cream screen and pulled the eye before the primary action beside
/// it — and the lime without the disc is the faintest thing on the screen. The shape alone says
/// whose logo it is.
struct PubkyMark: View {
    var size: CGFloat = 22

    var body: some View {
        Image("PubkyMark")
            .renderingMode(.template)
            .resizable()
            .scaledToFit()
            .frame(width: size, height: size)
            .foregroundStyle(LoopkyColor.foregroundSecondary)
    }
}

/// pubky.app's lime, and the halo the icon button below wears.
///
/// Deliberately *not* a `LoopkyColor` entry: it is another product's identity, and that palette is
/// Loopky's. It lives beside the mark it belongs to, and `PubkyAppLink.kt` holds the same values
/// for the same reason.
private enum PubkyBrand {
    /// `#C8FF00`, the `--brand` of pubky-app's own stylesheet.
    static let lime = UIColor(red: 200 / 255, green: 1, blue: 0, alpha: 1)

    /// One lime for both schemes — a glow is light, and light does not invert with the ground —
    /// but two alphas, because a shadow *composites* rather than adds: the lime at the light
    /// value over the dark ground lands on an olive that is no longer the brand colour, while the
    /// same value on cream already reads as lime.
    static let glowNear = glow(light: 0.70, dark: 0.85)
    static let glowFar = glow(light: 0.50, dark: 0.65)

    private static func glow(light: CGFloat, dark: CGFloat) -> Color {
        Color(UIColor { lime.withAlphaComponent($0.userInterfaceStyle == .dark ? dark : light) })
    }
}

private extension View {
    /// Two shadows rather than one, because SwiftUI's is a literal blur with no elevation model to
    /// lean on: the near one reads as the edge of the light, the far one is the halo. A single
    /// radius gives either a hard ring or a formless smudge. Neither is offset — nothing here is
    /// lit from above.
    func pubkyGlow() -> some View {
        shadow(color: PubkyBrand.glowNear, radius: 6)
            .shadow(color: PubkyBrand.glowFar, radius: 16)
    }
}

/// The button that leaves for pubky.app, in the same circle Share wears beside it — so it carries
/// no more weight in the row than that does — lit by ``PubkyBrand``'s halo.
///
/// The mark itself stays grey: the lime *as ink* is the faintest thing on a cream screen, 1.3:1,
/// and the dark plate that would fix that made this secondary control louder than the primary
/// action beside it. The brand colour rides underneath instead, where being pale costs it nothing.
struct PubkyAppIconButton: View {
    var action: () -> Void

    var body: some View {
        Button(action: action) {
            PubkyMark(size: 22)
                .frame(width: 44, height: 44)
                .background(Circle().fill(LoopkyColor.surfaceCard).pubkyGlow())
                .overlay(Circle().stroke(LoopkyColor.borderSubtle, lineWidth: 1))
        }
        .buttonStyle(.plain)
        // The mark has no text of its own, so the label goes on the button.
        .accessibilityLabel(Text("pubky_app_open_profile"))
        .accessibilityIdentifier("pubky_app_open_profile")
    }
}

/// The self-profile call to action: one soft row explaining what the button does, for the person
/// who has no reason to know that the key they signed in with is also a social account.
///
/// A card rather than a banner, and it never claims a Loopky deck appears there — it does not.
/// What travels is the profile and, when they choose to announce one, the post.
struct PubkyAppProfileCta: View {
    var action: () -> Void

    var body: some View {
        Button(action: action) {
            HStack(spacing: 14) {
                PubkyMark(size: 26)
                VStack(alignment: .leading, spacing: 2) {
                    Text("pubky_app_cta_title")
                        .font(.system(size: 14, weight: .bold))
                        .foregroundStyle(LoopkyColor.foregroundPrimary)
                    Text("pubky_app_cta_body")
                        .font(.system(size: 12))
                        .foregroundStyle(LoopkyColor.foregroundMuted)
                        .multilineTextAlignment(.leading)
                        .fixedSize(horizontal: false, vertical: true)
                }
                Spacer(minLength: 0)
                Image(systemName: "arrow.up.forward.square")
                    .font(.system(size: 16))
                    .foregroundStyle(LoopkyColor.foregroundMuted)
            }
            .frame(maxWidth: .infinity, alignment: .leading)
            .padding(16)
            .background(RoundedRectangle(cornerRadius: 20).fill(LoopkyColor.surfaceSecondary))
        }
        .buttonStyle(.plain)
        .accessibilityIdentifier("profile_pubky_app_cta")
    }
}
